### PR TITLE
[SPARK-47702][CORE] Remove Shuffle service endpoint from the locations list when RDD block is removed form a node

### PR DESCRIPTION
**Pull request by attilapiros**
### What changes were proposed in this pull request?

Credit to @maheshk114 for the initial investigation and the fix.

This PR fix a bug where the shuffle service's ID is kept among the block location list at the removing of a RDD block from a node. Before this change `StorageLevel.NONE` is used to notify about the block remove which causes the block manager master ignoring the update of the locations for shuffle service's IDs (for details please see the method `BlockManagerMasterEndpoint#updateBlockInfo()` and keep in mind `StorageLevel.NONE.useDisk` is `false`). But after this change only the replication count is set to 0 to notify the block remove so `StorageLevel#isValid` is still false but `storageLevel.useDisk` is kept as `true` this way the the shuffle service's ID will be removed from the block location list.

### Why are the changes needed?

If the block location is not updated properly, then tasks fails with fetch failed exception. The tasks will try to read the RDD blocks from a node using external shuffle service. The read will fail, if the node is already decommissioned.

```
WARN BlockManager [Executor task launch worker for task 25.0 in stage 6.0 (TID 1567)]: Failed to fetch remote block rdd_5_25 from BlockManagerId(4, vm-92303839, 7337, None) (failed attempt 1)
org.apache.spark.SparkException: Exception thrown in awaitResult: 
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:301)
	at org.apache.spark.network.BlockTransferService.fetchBlockSync(BlockTransferService.scala:103)
	at org.apache.spark.storage.BlockManager.fetchRemoteManagedBuffer(BlockManager.scala:1155)
	at org.apache.spark.storage.BlockManager.$anonfun$getRemoteBlock$8(BlockManager.scala:1099)
	at scala.Option.orElse(Option.scala:447)
	at org.apache.spark.storage.BlockManager.getRemoteBlock(BlockManager.scala:1099)
	at org.apache.spark.storage.BlockManager.getRemoteValues(BlockManager.scala:1045)
	at org.apache.spark.storage.BlockManager.get(BlockManager.scala:1264)
	at org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:1326)
```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added a new UT.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by attilapiros** [comment]
This is an updated version of https://github.com/apache/spark/pull/45836

**Comment by attilapiros** [comment]
cc @mridulm

**Comment by attilapiros** [comment]
cc @LuciferYang

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 1]
```diff
@@ -2112,8 +2112,10 @@ private[spark] class BlockManager(
       hasRemoveBlock = true
       if (tellMaster) {
         // Only update storage level from the captured block status before deleting, so that
-        // memory size and disk size are being kept for calculating delta.
-        reportBlockStatus(blockId, blockStatus.get.copy(storageLevel = StorageLevel.NONE))
+        // memory size and disk size are being kept for calculating delta. Reset the replica
+        // count 0 in storage level to notify that its a remove operation.
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala:2116]: typo? `its` -> `it's` ?

File: core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala [hunk 2]
```diff
@@ -100,12 +102,20 @@ trait BlockManagerReplicationBehavior extends SparkFunSuite
     // to make cached peers refresh frequently
     conf.set(STORAGE_CACHED_PEERS_TTL, 10)
 
+    conf.set(SHUFFLE_SERVICE_ENABLED, true)
+    conf.set(SHUFFLE_SERVICE_FETCH_RDD_ENABLED, true)
```
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:106]:
```
24/10/03 23:36:59.035 pool-1-thread-1-ScalaTest-running-BlockManagerReplicationSuite INFO BlockManager: external shuffle service port = 7337
24/10/03 23:36:59.174 pool-1-thread-1-ScalaTest-running-BlockManagerReplicationSuite INFO BlockManager: external shuffle service port = 7337
24/10/03 23:36:59.376 pool-1-thread-1-ScalaTest-running-BlockManagerReplicationSuite INFO BlockManager: external shuffle service port = 7337
24/10/03 23:36:59.476 pool-1-thread-1-ScalaTest-running-BlockManagerReplicationSuite INFO BlockManager: external shuffle service port = 7337
24/10/03 23:36:59.601 pool-1-thread-1-ScalaTest-running-BlockManagerReplicationSuite INFO BlockManager: external shuffle service port = 7337
24/10/03 23:36:59.648 pool-1-thread-1-ScalaTest-running-BlockManagerReplicationSuite INFO BlockManager: external shuffle service port = 7337
24/10/03 23:36:59.681 pool-1-thread-1-ScalaTest-running-BlockManagerReplicationSuite INFO BlockManager: external shuffle service port = 7337
24/10/03 23:37:00.657 pool-1-thread-1-ScalaTest-running-BlockManagerReplicationSuite INFO BlockManager: external shuffle service port = 7337
24/10/03 23:37:00.880 pool-1-thread-1-ScalaTest-running-BlockManagerReplicationSuite INFO BlockManager: external shuffle service port = 7337
24/10/03 23:37:00.935 pool-1-thread-1-ScalaTest-running-BlockManagerReplicationSuite INFO BlockManager: external shuffle service port = 7337
24/10/03 23:37:00.973 pool-1-thread-1-ScalaTest-running-BlockManagerReplicationSuite INFO BlockManager: external shuffle service port = 7337
```
When executing test cases that inherit from `BlockManagerReplicationBehavior`(such as `build/sbt clean "core/testOnly org.apache.spark.storage.BlockManagerBasicStrategyReplicationSuite"`), the logs will contain additional content as described above.

It seems that this signifies all subclasses of "BlockManagerReplicationBehavior" will be switched to using the external shuffle service during testing.

The affected test cases include:

- BlockManagerBasicStrategyReplicationSuite
- BlockManagerProactiveReplicationSuite
- BlockManagerReplicationSuite
- SslBlockManagerBasicStrategyReplicationSuite
- SslBlockManagerProactiveReplicationSuite
- SslBlockManagerReplicationSuite

Can we reduce this impact?

Comment (attilapiros) [reply core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala]: Addressed in https://github.com/apache/spark/pull/47779/commits/915d3e5404c1c315baae91c51575612af7148ddc

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Comment by LuciferYang** [comment]
Merged into master and branch-3.5. Thanks @attilapiros

**Comment by LuciferYang** [comment]
@attilapiros I have reverted the merge on branch-3.5 because it was causing compilation failures. If you have the time, please submit an independent pr for branch-3.5. Thanks ~

**Comment by attilapiros** [comment]
@LuciferYang 

Opened a new PR for branch-3.5:
https://github.com/apache/spark/pull/48357

**Pull request closed without merging**